pipeline_kinesis.c

```diff
@@ -0,0 +1,920 @@
+static void
+load_consumer_state(KinesisConsumerState *state, Oid oid)
+{
+	TupleTableSlot *slot;
+	int rv;
+
+	const char *query = "select e.*, c.* from pipeline_kinesis_consumers c "
+		"inner join pipeline_kinesis_endpoints e on c.endpoint = e.name "
+		"where c.oid = $1;";
+
+    Oid argtypes[1] = { OIDOID };
+	Datum values[1];
+    char nulls[1];
+
+	Datum d;
+	bool isnull = false;
+	MemoryContext old;
+
+	MemSet(state, 0, sizeof(KinesisConsumerState));
+
+	MemSet(nulls, 0, sizeof(nulls));
+	values[0] = ObjectIdGetDatum(oid);
+
+	SPI_connect();
+
+	rv = SPI_execute_with_args(query, 1, argtypes, values, nulls, false, 1);
+
+	if (rv != SPI_OK_SELECT)
+		elog(ERROR, "could not load consumer state %d", oid);
+
+	slot = MakeSingleTupleTableSlot(SPI_tuptable->tupdesc);
+	ExecStoreTuple(SPI_tuptable->vals[0], slot, InvalidBuffer, false);
+
+	old = MemoryContextSwitchTo(CacheMemoryContext);
+
+	state->id = oid;
+	state->endpoint_name =
+		TextDatumGetCString(slot_getattr(slot, 1, &isnull));
+	state->endpoint_region =
+		TextDatumGetCString(slot_getattr(slot, 2, &isnull));
+
+	d = slot_getattr(slot, 5, &isnull);
+
+	if (!isnull)
+		state->endpoint_credfile = TextDatumGetCString(d);
+
+	d = slot_getattr(slot, 5, &isnull);
+
+	if (!isnull)
+		state->endpoint_url = TextDatumGetCString(d);
+
+	state->kinesis_stream =
+		TextDatumGetCString(slot_getattr(slot, 6, &isnull));
+
+	state->relation =
+		TextDatumGetCString(slot_getattr(slot, 7, &isnull));
+
+	state->batchsize = DatumGetInt32(slot_getattr(slot, 8, &isnull));
+
+	MemoryContextSwitchTo(old);
+
+	SPI_finish();
+}
+
+/*
+ * find_shard_id
+ *
+ * Locate the index of a particular shard inside the stream metadata
+ */
+static int
+find_shard_id(kinesis_stream_metadata *meta, const char *id)
+{
+	int i = 0;
+	int num_shards = kinesis_stream_metadata_get_num_shards(meta);
+
+	for (i = 0; i < num_shards; ++i)
+	{
+		const kinesis_shard_metadata *smeta =
+			kinesis_stream_metadata_get_shard(meta, i);
+		const char *sid = kinesis_shard_metadata_get_id(smeta);
+
+		if (strcmp(sid, id) == 0)
+			return i;
+	}
+
+	return -1;
+}
+
+/*
+ * load_shard_state
+ *
+ * Given a partially setup consumer state and some stream metadata,
+ * initialize and load shard state (seqnums) from the kinesis tables
+ */
+static void
+load_shard_state(KinesisConsumerState *state, kinesis_stream_metadata *meta)
+{
+	const char *query = "select * from pipeline_kinesis_seqnums "
+		"where consumer_id = $1";
+
+	int num_shards = kinesis_stream_metadata_get_num_shards(meta);
+	int i = 0;
+	Oid argtypes[1] = { OIDOID };
+	Datum values[1];
+    char nulls[1];
+	int rv = 0;
+	TupleTableSlot *slot = NULL;
+
+	MemoryContext old = MemoryContextSwitchTo(CacheMemoryContext);
+	state->num_shards = num_shards;
+	state->shards = palloc0(num_shards * sizeof(ShardState));
+
+	for (i = 0; i < num_shards; ++i)
+	{
+		const kinesis_shard_metadata *smeta =
+			kinesis_stream_metadata_get_shard(meta, i);
+
+		state->shards[i].seqnum = makeStringInfo();
```

Comment:
This sequence number gets updated in the bgworker consume loop. No maximum length is specified for the sequence number anywhere. The underlying C++ type is a std::string.

Seems easier to use a stringinfo rather than palloc / repalloc